File: CPP/Human.cpp
// import library
#include <string>
#include <vector>
#include <iostream>
#include <iomanip>
#include <fstream>

// using standard namespace
using namespace std;

// deklarasi class Human. Best practicenya,
// nama class pakai huruf kapital depannya
class Human
{
private:
    // atribut privat
    int id;
    string name;
    string bidang;
    string partai;

public:
    // atribut publik
    // konstruktor
    Human()
    {
        // set default value untuk name dan gender
        this->name = "";
    }

    // konstruktor dengan parameter
    // bakal dipanggil kalo objek di-instansiasi dengan parameter
    Human(int id, string name, string bidang, string partai)
    {
        this->id = id;
        this->name = name;
        this->bidang = bidang;
        this->partai = partai;
    }

    /*
    Method get disesuaikan dengan tipe data yang bakal direturn nantinya
    */

    // get name
    int getID() const
    {
        return this->id;
    }

    // set name
    void setID(int id)
    {
        this->id = id;
    }

    // get name
    string getName() const
    {
        return this->name;
    }

    // set name
    void setName(string name)
    {
        this->name = name;
    }

    // get name
    string getBidang() const
    {
        return this->bidang;
    }

    // set name
    void setBidang(string bidang)
    {
        this->bidang = bidang;
    }

    // get name
    string getPartai() const
    {
        return this->partai;
    }

    // set name
    void setPartai(string partai)
    {
        this->partai = partai;
    }

    // method eat
    void eat()
    {
        cout << this->name << " is eating!" << '\n';
    }

    // method sleep
    void sleep()
    {
        cout << this->name << " is sleeping!" << '\n';
    }

    // method destruktor
    ~Human() {}
};

class HumanList
{
private:
    vector<Human> humans;

public:
    // Method untuk menambahkan Human baru ke dalam list
    void addHuman(const Human &human)
    {
        humans.push_back(human);
    }

    // Method untuk menghapus Human dari list berdasarkan ID
    void removeHumanByID(int id)
    {
        for (auto it = humans.begin(); it != humans.end(); ++it)
        {
            if (it->getID() == id)
            {
                humans.erase(it);
                return;
            }
        }
        cout << "Human with ID " << id << " not found." << endl;
    }

    // Method untuk mengedit data Human berdasarkan ID
    void editHumanByID(int id, const string &newName, const string &newBidang, const string &newPartai)
    {
        for (auto &human : humans)
        {
            if (human.getID() == id)
            {
                human.setName(newName);
                human.setBidang(newBidang);
                human.setPartai(newPartai);
                return;
            }
        }
        cout << "Human with ID " << id << " not found." << endl;
    }

    // Method untuk menampilkan seluruh data Human dalam list
    void displayAllHumans()
    {
        cout << "List of Humans:" << endl;
        for (const auto &human : humans)
        {
            cout << "ID: " << human.getID() << ", Name: " << human.getName() << ", Bidang: " << human.getBidang() << ", Partai: " << human.getPartai() << endl;
        }
    }

    // method destruktor
    ~HumanList() {}
};

File: Java/Human.java
// import library
import java.io.FileWriter;
import java.io.PrintWriter;

// deklarasi class Human. Best practice-nya,
// nama class pakai huruf kapital depannya
public class Human {
    // atribut privat
    private int id;
    private String name;
    private String bidang;
    private String partai;

    // konstruktor default
    public Human() {
        // set default value untuk name
        this.name = "";
    }

    // konstruktor dengan parameter
    // bakal dipanggil kalo objek di-instansiasi dengan parameter
    public Human(int id, String name, String bidang, String partai) {
        this.id = id;
        this.name = name;
        this.bidang = bidang;
        this.partai = partai;
    }

    // get id
    public int getID() {
        return this.id;
    }

    // set id
    public void setID(int id) {
        this.id = id;
    }

    // get name
    public String getName() {
        return this.name;
    }

    // set name
    public void setName(String name) {
        this.name = name;
    }

    // get bidang
    public String getBidang() {
        return this.bidang;
    }

    // set bidang
    public void setBidang(String bidang) {
        this.bidang = bidang;
    }

    // get partai
    public String getPartai() {
        return this.partai;
    }

    // set partai
    public void setPartai(String partai) {
        this.partai = partai;
    }

    // method eat
    public void eat() {
        System.out.println(this.name + " is eating!");
    }

    // method sleep
    public void sleep() {
        System.out.println(this.name + " is sleeping!");
    }

    // method destructor
    protected void finalize() {
        // not implemented, as Java handles garbage collection automatically
    }

    // method to write human details to file
    public void writeToFile(String filename) {
        try {
            FileWriter fileWriter = new FileWriter(filename, true);
            PrintWriter printWriter = new PrintWriter(fileWriter);
            printWriter.println("ID: " + id);
            printWriter.println("Name: " + name);
            printWriter.println("Bidang: " + bidang);
            printWriter.println("Partai: " + partai);
            printWriter.close();
        } catch (Exception e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        }
    }
}
